typescript/token/src/adapters/EvmTokenAdapter.ts

```diff
@@ -0,0 +1,213 @@
+import { BigNumber, PopulatedTransaction } from 'ethers';
+
+import {
+  BaseEvmAdapter,
+  ChainName,
+  MultiProtocolProvider,
+} from '@hyperlane-xyz/sdk';
+import {
+  Address,
+  Domain,
+  addressToByteHexString,
+  addressToBytes32,
+  bytes32ToAddress,
+  isValidAddressEvm,
+  strip0x,
+} from '@hyperlane-xyz/utils';
+
+import { MinimalTokenMetadata } from '../config';
+import {
+  ERC20,
+  ERC20__factory,
+  HypERC20,
+  HypERC20Collateral__factory,
+  HypERC20__factory,
+} from '../types';
+
+import {
+  IHypTokenAdapter,
+  ITokenAdapter,
+  TransferParams,
+  TransferRemoteParams,
+} from './ITokenAdapter';
+
+// Interacts with native currencies
+export class EvmNativeTokenAdapter
+  extends BaseEvmAdapter
+  implements ITokenAdapter
+{
+  async getBalance(address: Address): Promise<string> {
+    if (!isValidAddressEvm(address)) return '0';
+    const balance = await this.getProvider().getBalance(address);
+    return balance.toString();
+  }
+
+  async getMetadata(): Promise<MinimalTokenMetadata> {
+    // TODO get metadata from chainMetadata config
+    throw new Error('Metadata not available to native tokens');
+  }
+
+  async populateApproveTx(
+    _params: TransferParams,
+  ): Promise<PopulatedTransaction> {
+    throw new Error('Approve not required for native tokens');
```

Comment:
throw or noop?

Reply:
I slightly prefer throw because returning undefined would require us to change the signature across adapters to be `Promise<PopulatedTransaction | undefined>`